### PR TITLE
Introduce course selection limit for continuous applications

### PR DESCRIPTION
**Pull request by JR-G**
## Context

Adding a limit to the number of applications in draft (pre submission) and in progress (post submission) that a candidate can add. The limit is 4 (total) which is the current application choice cap.

## Changes proposed in this pull request
- Render copy and hide the 'add application' button if the candidate has 4 draft/submitted applications
- Redirect away from the add a course journey is this limit is true
- Pluralise the 'You can add X application(s)' count
- Ensure the detect invariant check will fire correctly in continuous applications (i.e. taking into consideration inactive)

|Can add a choice|Can no longer add a choice|
|---|---|
|image|image|

## Guidance to review

- Play around with a different combination of application choices. Try and manually enter URLs into the browser to ensure the redirects work
- Should we raise an error on the `SubmitApplicationChoice` service if `can_add_more_choices?` returns `false`?

## Link to Trello card

https://trello.com/c/PKgtbpJ2/446-ca-course-selection-limit-and-validate-maximum-applications-in-draft-in-progress

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: spec/workers/detect_invariants_daily_check_spec.rb [hunk 1]
```diff
@@ -74,29 +74,61 @@
...
-      allow(Sentry).to receive(:capture_exception).with(an_instance_of(described_class::SubmittedApplicationHasMoreThanTheMaxCourseChoices))
+    context 'when continuous application', continuous_applications: true do
+      it 'detects submitted applications with more than the maximum number of course choices' do
+        allow(Sentry).to receive(:capture_exception).with(an_instance_of(described_class::SubmittedApplicationHasMoreThanTheMaxCourseChoices))
 
-      good_application_form = create(:completed_application_form, submitted_application_choices_count: 4)
-      bad_application_form = create(:completed_application_form, submitted_application_choices_count: 3)
-      bad_application_form.application_choices << build_list(:application_choice, 2, status: :offer)
-      ApplicationChoice.all.each { |a| a.update_course_option_and_associated_fields! create(:course_option) }
+        good_application_form = create(:completed_application_form, submitted_application_choices_count: 4)
```
Comment (tomas-stefano) [inline spec/workers/detect_invariants_daily_check_spec.rb:81]: Question: Does the invariant check counts the scenario when the user could have 10 submitted applications (6 inactive and 4 awaiting for example)?

**Comment by tomas-stefano** [comment]
> Should we raise an error on the SubmitApplicationChoice service if can_add_more_choices? returns false?

I think we should add a validation on app/services/candidate_interface/continuous_applications/application_choice_submission.rb for this, for security purposes.

**Review by tomas-stefano — APPROVED** [review]
2 Minor considerations but all looks awesome 🚀

File: app/views/candidate_interface/continuous_applications_choices/index.html.erb [hunk 2]
```diff
@@ -6,19 +6,27 @@
       <%= t('page_titles.continuous_applications.your_applications') %>
     </h1>
 
-    <%= render CandidateInterface::ApplicationsLeftMessageComponent.new(@application_form_presenter.application_form) %>
+    <% unless @application_form_presenter.can_add_more_choices? %>
+      <p class="govuk-body">You cannot add any more applications.</p>
+      <p class="govuk-body">If one of your applications is unsuccessful, or you withdraw it, you will be able to add another application.<p>
```
Comment (MaeveRoseveare) [inline app/views/candidate_interface/continuous_applications_choices/index.html.erb:11]:
```suggestion
      <p class="govuk-body">If one of your applications is unsuccessful, or you withdraw or remove it, you will be able to add another application.<p>
```

**Review by MaeveRoseveare — APPROVED** [review]
Looks good, just added a bit to account for any draft applications, in which case they would need to remove it.

For Tomas' comment on needing validation - do we need content? I'm not sure I understand the issue correctly, but if you need content let me know.

File: spec/workers/detect_invariants_daily_check_spec.rb (hunk 1 above)
Comment (JR-G) [reply spec/workers/detect_invariants_daily_check_spec.rb:81]: good q! – will add

**Comment by tomas-stefano** [comment]
> Looks good, just added a bit to account for any draft applications, in which case they would need to remove it.
> 
> For Tomas' comment on needing validation - do we need content? I'm not sure I understand the issue correctly, but if you need content let me know.

Not necessary any content but imagine the following:

1. Candidate apply 4 times
2. Candidate (that knows about how HTTP protocol works)
3. Do a POST request to /submit url
4. Because we don't validate the candidate applied more than 4 times! (although the invariant check will let us know but better be safe!)

**Comment by JR-G** [comment]
> > Looks good, just added a bit to account for any draft applications, in which case they would need to remove it.
> > For Tomas' comment on needing validation - do we need content? I'm not sure I understand the issue correctly, but if you need content let me know.
> 
> Not necessary any content but imagine the following:
> 
> 1. Candidate apply 4 times
> 2. Candidate (that knows about how HTTP protocol works)
> 3. Do a POST request to /submit url
> 4. Because we don't validate, then the candidate would applied more than 4 times! (although the invariant check will let us know but better be safe!)

It also would appear if the candidate managed to _somehow_ get on the submission page (pls excuse old capybara screenshot):
image

**Pull request merged**